accepted/2022/version-selection.md

```diff
@@ -0,0 +1,447 @@
+
+Re-compiling applications to run on a new .NET version is already possible, while the configuration option is not.
+
+Note: It is possible and likely that some applications will fail to run on a new .NET version. That's OK. Imagine 80% of apps just work and 20% do not. That means that you now have a much smaller group of apps to address, and can have a bit more time to do so. Admittedly, you'd need some partitioning in your environment to deploy such a policy.
+
+## Principles
+
+Version selection should have the following principles:
+
+- The `dotnet` that is launched is the one that is used, with no magic policies by default.
+- Convenience/opinionated/magic experiences are opt-in.
+- Common tasks should be made easy/easier, possibly via opt-in convenience experiences.
+- SDK and runtime selection policies and controls should be aligned, as much as possible (to aid intuition and overall simplicity).
+
+## Model
+
+Let's briefly recap the fundamentals of the version selection model. It is based on the following configuration concepts:
+
+- **Install root** -- Where should the host look for .NET versions?
+- Which version should be selected?
+  - **Minimum version** -- What's the minimum version that is required?
+  - **Roll forward policy** -- What's the policy for determining the highest version acceptable?
+  - **Preview versions** -- Are preview versions acceptable to use?
+- **Configuration** -- Which configuration knobs are available for answering these questions?
+
+.NET hosts select versions for two components:
+
+- Runtimes (AKA "frameworks", under the `shared` directory)
+- SDKs (under the `sdk` directory)
+
+Note: Apphost (`myapp`) and custom hosts only select runtimes. The `dotnet` host is responsible for selecting both SDKs and runtimes.
+
+The following sections describe the updates in detail. They are intended be satisfy the customer scenarios described earlier (at least, in part). They also propose improvements to known gaps and bugs that are distinct from the customer scenarios. Existing behavior is also provided as a baseline to help readers understand how the proposed changes fit in.
+
+## Location configuration
+
+.NET hosts expose the following (existing and proposed) location configuration for discovering the .NET install root location.
+
+For SDKs and runtimes:
+
+- `PATH` -- The use of `dotnet` is resolved by the `PATH`.
+- Absolute path -- The use of an absolute path to `dotnet` (like `C:\PrivateSDK\dotnet`).
+- [NEW SDK scenario] Convention location -- `dotnet` looks for SDKs and runtimes in the `.dotnet` directory in the current or (grand-*)parent directory.
```

Comment:
Using current directory or its parents to look for runtimes looks very problematic from the security point of view. It means that if somebody manages to drop a file in .dotnet subdirectory somewhere on your machine, `dotnet myapp.dll` will start executing.

Reply:
That is covered later in the spec. We're proposing the same model as Visual Studio Code Trusted Workspaces.